File: C++/25. Reverse Nodes in k-Group.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(NULL) {}
 *     ListNode(int x) : val(x), next(NULL) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(NULL) {}
 *     ListNode(int x) : val(x), next(NULL) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* reverseKGroup(ListNode* head, int k) {
        if(head==NULL || head->next==NULL || k == 1) {return head;}

        ListNode *curr = head;
        ListNode *sol = NULL;
        ListNode *dum = NULL;

        while(curr!=NULL) {
            ListNode *l = curr;
            ListNode *r = NULL;
            int count = 0;
            while(curr!=NULL && count != k) {
                r = curr;
                curr=curr->next;
                count++;
            }

            if(count != k) {
                if(sol==NULL) {return head;}
                dum->next = l;
                return sol;
            }

            r->next = NULL;
            ListNode *rev = reverse(l);

            if(sol==NULL) {sol = rev;}
            else {dum->next = rev;}
            dum = l;
        }
        return sol;
    }
    ListNode* reverse(ListNode *curr) {
        ListNode *next = curr, *prev = NULL;
        while(curr) {
            next =curr->next;
            curr->next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
};

File: C#/25. Reverse Nodes in k-Group.cs
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public int val;
 *     public ListNode next;
 *     public ListNode(int val=0, ListNode next=null) {
 *         this.val = val;
 *         this.next = next;
 *     }
 * }
 */
public class Solution {
    public ListNode ReverseKGroup(ListNode head, int k) {
        if(head==null || head.next==null || k == 1) {return head;}

        ListNode curr = head;
        ListNode sol = null;
        ListNode dum = null;

        while(curr!=null) {
            ListNode l = curr;
            ListNode r = null;
            int count = 0;
            while(curr!=null && count != k) {
                r = curr;
                curr = curr.next;
                count++;
            }

            if(count != k) {
                if(sol==null) {return head;}
                dum.next = l;
                return sol;
            }

            r.next = null;
            ListNode rev = reverse(l);

            if(sol==null) {sol = rev;}
            else {dum.next = rev;}
            dum = l;
        }
        return sol;
    }
    public ListNode reverse(ListNode curr) {
        ListNode next = curr, prev = null;
        while(curr!=null) {
            next = curr.next;
            curr.next = prev;
            prev = curr;
            curr = next;
        }
        return prev;
    }
}
